packages/frint-router/README.md

```diff
@@ -0,0 +1,239 @@
+  - [HashRouterService](#hashrouterservice)
+  - [MemoryRouterService](#memoryrouterservice)
+
+<!-- /MarkdownTOC -->
+
+---
+
+# Guide
+
+## Installation
+
+With [npm](https://www.npmjs.com/):
+
+```
+$ npm install --save frint-router
+```
+
+Via [unpkg](https://unpkg.com) CDN:
+
+```html
+<script src="https://unpkg.com/frint-router@latest/dist/frint-router.min.js"></script>
+
+<script>
+  // available as `window.FrintRouter`
+</script>
+```
+
+## Usage
+
+The classes exported by this package are all independent and can be used as is.
+
+```js
+import HashRouterService from 'frint-router/HashRouterService';
+// import BrowserRouterService from 'frint-router/BrowserRouterService';
+// import MemoryRouterService from 'frint-router/MemoryRouterService';
+
+const router = new HashRouterService();
+const subscription = router.getHistory$().subscribe(function (history) {
+  console.log(history);
+  // {
+  //   length: 1,
+  //   location: { ... }, // Object liked `window.location`
+  //   action: 'PUSH',
+  // }
+});
+```
+
+The subscription will keep emitting new values every time there is a change in history.
+
+### Convention
+
+To connect it well with other packages, we need to follow a convention of using the provider name `router`:
+
+```js
+import { createApp } from 'frint';
+import HashRouterService from 'frint-router/HashRouterService';
+
+const RootApp = createApp({
+  name: 'MyRootApp',
+  providers: [
+    {
+      name: 'router',
+      useFactory: function () {
+        return new HashRouterService();
+      },
+      cascade: true,
+    }
+  ],
+});
+```
+
+### Direct imports
+
+It is advised to import the appropriate router service class directly from the package, to make sure you are only bundling the services you explicitly need.
+
+```js
+// will bundle only the individual HashRouterService
+import HashRouterService from 'frint-router/HashRouterService';
+
+// will bring ALL the service classes
+import { HashRouterService } from 'frint-router';
+```
+
+---
+
+# API
+
+The package exports three classes:
+
+* `BrowserRouterService`: uses HTML5 History API
+* `HashRouterService`: For legacy browsers
+* `MemoryRouterService`: useful for tests
+
+All of them implement the same set of methods.
+
+## BrowserRouterService
+
+### getHistory$
+
+> getHistory$()
+
+Streams history as it changes over time.
+
+#### Returns
+
+`Observable`: Streams `history` Object.
+
+Structure of `history` object:
+
+```js
+{
+  length: 1, // number of entries
+  location: { ... }, // like `window.location`
+  action: 'PUSH' // either PUSH, REPLACE, or POP
+}
+```
+
+### getMatch$
+
+> getMatch$(pattern, options)
+
+Keeps matching pattern against history as it keeps changing over time.
+
+#### Arguments
+
+1. `pattern` (`String`): Pattern to match against
+1. `options` (`Object`)
+  * `options.exact` (`Boolean`): Matches pattern exactly, defaults to `false`.
+  * `options.cache` (`Boolean`): Cache matches (TODO)
```

Comment:
a minimal amount of caching should be supported. repeated route matching can be costly on component re-renders. I will take care of it.